### PR TITLE
Add support for basic auth

### PR DESCRIPTION
**Pull request by bion**

### CODE REVIEW

**Comment by sethvargo** [comment]
Hey @bion 

Thank you for the PR. This is looking really good!

What do you think about making this a single option (`--auth` or `--basic-auth`) that accepts `USERNAME[:PASSWORD]` and we can split on the first colon on the backend. I think this is more closely how curl behaves, so I would like to mirror that API as closely as possible. What do you think?

/cc @ryanuber

**Comment by ryanuber** [comment]
@bion I agree with @sethvargo, it would be nice to have these as `user[:pass]` on the command line. I vote for the `-auth` flag.

File: cli.go [hunk 1]
```diff
@@ -279,6 +283,10 @@ func bootstrap(config *Config, dry bool, once bool) (*Runner, *watch.Watcher, er
 			},
 		}
 	}
+	if (config.HttpAuth.Username != "") || (config.HttpAuth.Password != "") {
```
Comment (ryanuber) [inline cli.go]: Probably don't even need to do this if check here.

File: config.go [hunk 2]
```diff
@@ -32,6 +37,8 @@ type Config struct {
 	// Consul. This requires Consul to be configured to serve HTTPS.
 	SSL bool `mapstructure:"ssl"`
 
+	HttpAuth HttpAuth `mapstructure:"HttpAuth"`
```
Comment (ryanuber) [inline config.go]: The `mapstructure` tag maps the value to the key in a config file, which should follow the snake-chaining (lowercase with underscores) convention.

Comment (sethvargo) [inline config.go]: But if we follow the "single-value" for authentication, this would likely just become a string in the config.

File: config.go [hunk 3]
```diff
@@ -65,6 +72,14 @@ func (c *Config) Merge(config *Config) {
 		c.SSLNoVerify = true
 	}
 
+	if (config.HttpAuth.Username != "") || (config.HttpAuth.Password != "") {
+		c.HttpAuth = config.HttpAuth
+	}
+
+	if config.HttpAuth.Password != "" {
```
Comment (ryanuber) [inline config.go]: This should probably be all-or-nothing for the best clarity. If user or password are present, then merge, else do nothing.

File: config_test.go [hunk 4]
```diff
@@ -119,6 +119,30 @@ func TestMerge_HttpsOptions(t *testing.T) {
 	}
 }
 
+func TestMerge_BasicAuthOptions(t *testing.T) {
+	{
```
Comment (ryanuber) [inline config_test.go]: Nitpick: We don't need to wrap this test in a block since the variables aren't used in any other scope.

**Comment by ryanuber** [comment]
@bion I added a few comments to the code as well. Thanks for doing this!

**Comment by bion** [comment]
Thanks for the comments, guys. This is my first foray into Go so please forgive any naivete. I'll address these comments and @ you when it's ready for another look.

**Comment by bion** [comment]
@ryanuber @sethvargo back to you guys.

**Comment by sethvargo** [comment]
Hey @bion 

Thank you for making those changes. I am working on some internal refactorings for Consul Template and I will merge this after those have solidified.

**Comment by sethvargo** [comment]
Cleaned up and merged in 9e819cc. Thanks!

**Pull request closed without merging**